content/posts/messaging-infrastructure.md

```diff
@@ -0,0 +1,80 @@
+
+Atomic store-send-and-consume is a very powerful concept as it allows writing very simple business logic code. Provided that this code does not produce any side effects, the behavior of the system is exactly as if *exactly-once message delivery* was a thing. Sounds compelling, doesn't it?So how can I use this mode? The atomic store-send-and-consume can be implemented in two ways. 
+
+First approach to use distributed transaction technology that ensures that individual transactions against the data store and the message queue are made atomic and durable. The problem is, such technology is no longer widely accessible. Few message queue products support distributed transactions and none of them is designed to work well in the cloud.
+
+The second approach is to use a single technology for both storing data and sending messages or in other words, emulate a message queue in a database. The drawback of this approach is that databases usually are not designed with the type of workload that is characteristic to message queues, leading to poor performance compared to dedicated message queue solutions.
+
+In summary, atomic store-send-and-consume is a great model from the simplicity of business logic perspective. Despite challenges related to technology, it should be always seriously considered when designing a distributed system.
+
+## Atomic store-and-send
+
+A commonly used middle ground between the two approaches described earlier is a mode called *atomic store-and-send* or *atomic store-and-publish*. In this mode there are two transactions
+- persist state and send out messages
+- consume incoming message
+As a consequence, there is one partial failure mode possible when the state has been persisted and messages sent out but the incoming message has not been consumed. To handle that case the code of the service needs to be able to check if a given message has already been processed. 
+
+You might now ask the question how can we implement a transaction that spans both persisting state and sending out messages. Wouldn't it require the same sort of infrastructure as the store-send-and-consume mode? As it turns out, it does not.
+
+If the case of store-send-and-consume we are dealing with a problem that is equivalent to the distributed consensus problem. We need two parties (the database and the queue) to agree with the relation to committing the state changes and consuming the message. This category of problems is proved to not have a solution in an *asynchronous* network (one with no guarantees for eventual message delivery). In a more realistic model of *partially synchronous network* the problem of consensus fortunately does have solutions but they are both difficult and costly.
```

Comment:
I think we may lose the reader here :)

Reply:
I'll remove the (a)synchronous network part.